Overload resolution error with indexers across assemblies

Repro code

```
    // lib1.cs 
    public abstract class TaskInput<TKey, TValue>
        where TKey : class
        where TValue : class
    {
        public virtual TValue this[TKey key] => throw null;
    }

    // lib2.cs
    public class CallNode { }
    public sealed class CalleeCallNodeInput<TValue> : TaskInput<CallNode, TValue>
        where TValue : class
    {
        public override TValue this[CallNode key] => throw null;
    }

    // data.cs
    class FunctionSummary { }

    class Repro
    {
        void M()
        {
            CalleeCallNodeInput<FunctionSummary> c = null;
            CallNode node = null;
            var summary = c[node];
        }

        static void Main(){ }
    }
```

Repro steps:

```
> csc lib1.cs /t:library
> csc lib2.cs /t:library /r:lib1.dll
> csc data.cs /r:lib1.dll /r:lib2.dll


data.cs(10,27): error CS0121: The call is ambiguous between the following methods or properties: 'TaskInput<TKey, TValue>.this[TKey]' and 'TaskInput<TKey, TValue>.this[TKey]'
```

This should compile without error or warnings. Likely this is a regession from covariant returns.

--- a/src/Compilers/CSharp/Portable/Symbols/EventSymbol.cs
+++ b/src/Compilers/CSharp/Portable/Symbols/EventSymbol.cs
@@ -168,8 +168,7 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
 
         internal EventSymbol GetLeastOverriddenEvent(NamedTypeSymbol? accessingTypeOpt)
         {
-            var accessingType = ((object?)accessingTypeOpt == null ? this.ContainingType : accessingTypeOpt).OriginalDefinition;
-
+            accessingTypeOpt = accessingTypeOpt?.OriginalDefinition;
             EventSymbol e = this;
             while (e.IsOverride && !e.HidesBaseEventsByName)
             {
@@ -195,7 +194,8 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
                 // See InternalsVisibleToAndStrongNameTests: IvtVirtualCall1, IvtVirtualCall2, IvtVirtual_ParamsAndDynamic.
                 EventSymbol? overridden = e.OverriddenEvent;
                 HashSet<DiagnosticInfo>? useSiteDiagnostics = null;
-                if ((object?)overridden == null || !AccessCheck.IsSymbolAccessible(overridden, accessingType, ref useSiteDiagnostics))
+                if ((object?)overridden == null ||
+                    (accessingTypeOpt is { } && !AccessCheck.IsSymbolAccessible(overridden, accessingTypeOpt, ref useSiteDiagnostics)))
                 {
                     break;
                 }

--- a/src/Compilers/CSharp/Portable/Symbols/PropertySymbol.cs
+++ b/src/Compilers/CSharp/Portable/Symbols/PropertySymbol.cs
@@ -253,8 +253,7 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
 
         internal PropertySymbol GetLeastOverriddenProperty(NamedTypeSymbol accessingTypeOpt)
         {
-            var accessingType = ((object)accessingTypeOpt == null ? this.ContainingType : accessingTypeOpt).OriginalDefinition;
-
+            accessingTypeOpt = accessingTypeOpt?.OriginalDefinition;
             PropertySymbol p = this;
             while (p.IsOverride && !p.HidesBasePropertiesByName)
             {
@@ -280,7 +279,8 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
                 // See InternalsVisibleToAndStrongNameTests: IvtVirtualCall1, IvtVirtualCall2, IvtVirtual_ParamsAndDynamic.
                 PropertySymbol overridden = p.OverriddenProperty;
                 HashSet<DiagnosticInfo> useSiteDiagnostics = null;
-                if ((object)overridden == null || !AccessCheck.IsSymbolAccessible(overridden, accessingType, ref useSiteDiagnostics))
+                if ((object)overridden == null ||
+                    (accessingTypeOpt is { } && !AccessCheck.IsSymbolAccessible(overridden, accessingTypeOpt, ref useSiteDiagnostics)))
                 {
                     break;
                 }

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionTests.cs
@@ -11263,5 +11263,74 @@ public static class Extensions
 
             CompileAndVerify(code, expectedOutput: @"2");
         }
+
+        [Fact]
+        public void GenericTypeOverriddenMethod()
+        {
+            var source0 =
+@"public class Base<TKey, TValue>
+    where TKey : class
+    where TValue : class
+{
+    public virtual TValue F(TKey key) => throw null;
+}";
+            var ref0 = CreateCompilation(source0).EmitToImageReference();
+
+            var source1 =
+@"public class A { }
+public class Derived<TValue> : Base<A, TValue>
+    where TValue : class
+{
+    public override TValue F(A key) => throw null;
+}";
+            var ref1 = CreateCompilation(source1, references: new[] { ref0 }).EmitToImageReference();
+
+            var source2 =
+@"class B { }
+class Program
+{
+    static void M(Derived<B> d, A a)
+    {
+        _ = d.F(a);
+    }
+}";
+            var comp = CreateCompilation(source2, references: new[] { ref0, ref1 });
+            comp.VerifyEmitDiagnostics();
+        }
+
+        [Fact]
+        [WorkItem(46549, "https://github.com/dotnet/roslyn/issues/46549")]
+        public void GenericTypeOverriddenProperty()
+        {
+            var source0 =
+@"public class Base<TKey, TValue>
+    where TKey : class
+    where TValue : class
+{
+    public virtual TValue this[TKey key] => throw null;
+}";
+            var ref0 = CreateCompilation(source0).EmitToImageReference();
+
+            var source1 =
+@"public class A { }
+public class Derived<TValue> : Base<A, TValue>
+    where TValue : class
+{
+    public override TValue this[A key] => throw null;
+}";
+            var ref1 = CreateCompilation(source1, references: new[] { ref0 }).EmitToImageReference();
+
+            var source2 =
+@"class B { }
+class Program
+{
+    static void M(Derived<B> d, A a)
+    {
+        _ = d[a];
+    }
+}";
+            var comp = CreateCompilation(source2, references: new[] { ref0, ref1 });
+            comp.VerifyEmitDiagnostics();
+        }
     }
 }

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionTests.cs
@@ -11274,8 +11274,6 @@ public static class Extensions
 {
     public virtual TValue F(TKey key) => throw null;
 }";
-            var ref0 = CreateCompilation(source0).EmitToImageReference();
-
             var source1 =
 @"public class A { }
 public class Derived<TValue> : Base<A, TValue>
@@ -11283,8 +11281,6 @@ public class Derived<TValue> : Base<A, TValue>
 {
     public override TValue F(A key) => throw null;
 }";
-            var ref1 = CreateCompilation(source1, references: new[] { ref0 }).EmitToImageReference();
-
             var source2 =
 @"class B { }
 class Program
@@ -11294,7 +11290,13 @@ class Program
         _ = d.F(a);
     }
 }";
-            var comp = CreateCompilation(source2, references: new[] { ref0, ref1 });
+
+            var comp = CreateCompilation(new[] { source0, source1, source2 });
+            comp.VerifyEmitDiagnostics();
+
+            var ref0 = CreateCompilation(source0).EmitToImageReference();
+            var ref1 = CreateCompilation(source1, references: new[] { ref0 }).EmitToImageReference();
+            comp = CreateCompilation(source2, references: new[] { ref0, ref1 });
             comp.VerifyEmitDiagnostics();
         }
 
@@ -11309,8 +11311,6 @@ class Program
 {
     public virtual TValue this[TKey key] => throw null;
 }";
-            var ref0 = CreateCompilation(source0).EmitToImageReference();
-
             var source1 =
 @"public class A { }
 public class Derived<TValue> : Base<A, TValue>
@@ -11318,8 +11318,6 @@ public class Derived<TValue> : Base<A, TValue>
 {
     public override TValue this[A key] => throw null;
 }";
-            var ref1 = CreateCompilation(source1, references: new[] { ref0 }).EmitToImageReference();
-
             var source2 =
 @"class B { }
 class Program
@@ -11329,7 +11327,13 @@ class Program
         _ = d[a];
     }
 }";
-            var comp = CreateCompilation(source2, references: new[] { ref0, ref1 });
+
+            var comp = CreateCompilation(new[] { source0, source1, source2 });
+            comp.VerifyEmitDiagnostics();
+
+            var ref0 = CreateCompilation(source0).EmitToImageReference();
+            var ref1 = CreateCompilation(source1, references: new[] { ref0 }).EmitToImageReference();
+            comp = CreateCompilation(source2, references: new[] { ref0, ref1 });
             comp.VerifyEmitDiagnostics();
         }
     }

